derive/src/pyclass.rs

```diff
@@ -316,11 +318,111 @@ pub fn impl_pyclass(attr: AttributeArgs, item: Item) -> Result<TokenStream2, Dia
+                let method_ident = Ident::new(&method_name, field_name.span());
+
+                let method = quote! {
+                    fn #method_ident(
+                        zelf: &::rustpython_vm::obj::objtuple::PyTuple,
+                        _vm: &::rustpython_vm::vm::VirtualMachine)
+                        -> ::rustpython_vm::pyobject::PyObjectRef {
+                        zelf.fast_getitem(#idx)
+                    }
+                };
+                methods.push(method);
+                let field_name_str = field_name.to_string();
+                // TODO add doc to the generated property
+                let method_reference = quote! {
+                    class.set_str_attr(
+                        #field_name_str,
+                        ::rustpython_vm::obj::objproperty::PropertyBuilder::new(ctx)
+                            .add_getter(Self::#method_ident)
+                            .create(),
+                    );
+                };
+                method_references.push(method_reference);
+                field_names.push(quote!(#field_name));
+            } else {
+                field_names.push(quote!(#idx));
+            }
+        }
+
+        let ty = &struc.ident;
+        let ret = quote! {
+            #struc
+            #class_def
+            impl #ty {
+                fn into_struct_sequence(&self,
+                    vm: &::rustpython_vm::vm::VirtualMachine,
+                    cls: ::rustpython_vm::obj::objtype::PyClassRef,
+                ) -> ::rustpython_vm::pyobject::PyResult<::rustpython_vm::obj::objtuple::PyTupleRef> {
+                    let tuple: ::rustpython_vm::obj::objtuple::PyTuple =
+                        vec![#(::rustpython_vm::pyobject::IntoPyObject
+                                    ::into_pyobject(self.#field_names, vm)?
+                             ),*].into();
+                    ::rustpython_vm::pyobject::PyValue::into_ref_with_type(tuple, vm, cls)
+                }
+
+                #(#methods)*
+            }
+            impl ::rustpython_vm::pyobject::PyClassImpl for #ty {
+                fn impl_extend_class(
+                    ctx: &::rustpython_vm::pyobject::PyContext,
+                    class: &::rustpython_vm::obj::objtype::PyClassRef,
+                ) {
+                    #(#method_references)*
+                }
+
+                fn make_class(
+                    ctx: &::rustpython_vm::pyobject::PyContext
+                ) -> ::rustpython_vm::obj::objtype::PyClassRef {
+                    let py_class = ctx.new_class(<Self as ::rustpython_vm::pyobject::PyClassDef>::NAME, ctx.tuple_type());
+                    Self::extend_class(ctx, &py_class);
+                    py_class
+                }
+            }
+        };
+        Ok(ret)
+    } else {
+        bail_span!(
+            item,
+            "#[pystruct_sequence] can only be on a struct declaration"
+        )
+    }
```

Comment:
If you want, I'd say you could put `let struc = if let ... { struc } else { bail_span!(...) };` at the beginning, to help keep the bulk of the code less indented.